modules/renter/upload.go

```diff
@@ -150,15 +151,54 @@ func (r *Renter) Upload(up modules.FileUploadParams) error {
 	r.save()
 	r.mu.Unlock(lockID)
 
-	// Upload a piece to every host on the network.
+	// Choose 3*redundancy hosts and upload a piece to 'redundancy' of them.
+	// This allows a significant portion of the hostdb to be outdated.
+	hostPool := r.hostDB.RandomHosts(3 * redundancy)
+	piecePool := f.Pieces
+	poolClosed := false
+	poolMutex := sync.New(250*time.Millisecond, 1)
 	errChan := make(chan error, len(f.Pieces))
-	for i := range f.Pieces {
-		// threadedUploadPiece will change the memory that the piece points
-		// to, which is useful because it means the file can be renamed
-		// without disrupting the upload process.
-		go func(piece *filePiece) {
-			errChan <- r.threadedUploadPiece(up, piece)
-		}(&f.Pieces[i])
+	for i := 0; i < parallelUploads; i++ {
+		go func() {
+			lockID := poolMutex.Lock()
+			pieceFinished := true
+			var piece *filePiece
+			poolMutex.Unlock(lockID)
+			for {
+				lockID := poolMutex.Lock()
+				if len(hostPool) == 0 || len(piecePool) == 0 {
+					if !pieceFinished {
+						errChan <- errors.New("upload for piece failed")
+					}
+					poolMutex.Unlock(lockID)
+					return
+				}
+				if pieceFinished {
+					piece = &piecePool[0]
+					piecePool = piecePool[1:]
+					pieceFinished = false
+				}
+				host := hostPool[0]
+				hostPool = hostPool[1:]
+				poolMutex.Unlock(lockID)
+
+				err := r.threadedUploadPiece(host, up, piece)
```

Comment:
this seems to require 1 host per piece. What happens if `len(hostPool)` < `len(piecePool)`?